### PR TITLE
Implemented download_pdb_in_range()

### PR DESCRIPTION
**Pull request by EricFlorin**
`download_pdb_in_range()` downloads and writes PDB files, whose PDB IDs are within a given range [lower, upper].

`download_pdb_in_range()` takes three inputs: `lower`, `upper`, and `dir`:
- `lower` and `upper` are strings that represents the range of IDs to look for PDB files. The range is from `lower` to `upper` inclusive. `lower` and `upper` can take in IDs with the wildcard character '*'; for example, if a user defines a range as "1A*" to "2*", then the function would attempt to download all PDB files with IDs between 1A00 to 2000 inclusive.
- `dir` is a string that represents a directory for the PDB files to be written to.

### CODE REVIEW

File: DeepProjection/pdbGenerator.py [hunk 1]
```diff
@@ -36,36 +36,176 @@ def get_random_pdb(self):
                 pass
 
     
-    def download_all_existing_pdb(self):
-        # Downloads all PDB files from the Protein Data Bank
+    def download_pdb_in_range(self, lower, upper, dir):
```
Comment (chuckie82) [inline DeepProjection/pdbGenerator.py]: May be useful to assign default values: upper=None and dir=None
If upper is None, then just use lower to determine what to download.
If dir is None, download in current working directory

File: DeepProjection/pdbGenerator.py [hunk 2]
```diff
@@ -36,36 +36,176 @@ def get_random_pdb(self):
...
 
-        # Time to generate the codes
-        # The first position is always a number from 0 - 9
-        for first_position in range (10):
+        
+        # Checking for Special Ranges
+        # If lower == upper, our loop code should be able to handle that case on its own.
+
+        if lower == '*' and upper == '*': # Special Range 2: lower == '*' and upper == '*', so we will download all files
+            lower = '0000'
```
Comment (chuckie82) [inline DeepProjection/pdbGenerator.py]: First character in the PDB ID is between 1 and 9. Not zero.
In other characters, zero to 9.

File: DeepProjection/pdbGenerator.py [hunk 3]
```diff
@@ -36,36 +36,176 @@ def get_random_pdb(self):
...
+                                print("download_pdb_in_range trying:", currentID)
+                                
+                                pdb = pypdb.get_pdb_file(currentID, filetype='pdb', compression=False)
+                                
+                                with open(os.path.join(dir, currentID + ".pdb") ,"w") as f:
+                                    f.writelines(pdb)
 
-        # For debug
-        print(len(list_of_valid_ids))
+                                print("downloaded: ", currentID)
```
Comment (chuckie82) [inline DeepProjection/pdbGenerator.py]: remove print statements

File: DeepProjection/pdbGenerator.py [hunk 4]
```diff
@@ -36,36 +36,176 @@ def get_random_pdb(self):
...
+        
+        upper : str
+            A PDB ID that will be the upper bound of our search.
+            The upper range input will be considered a "wildcard" if the string contains
+            an '*' after a value (e.g. 1A*) or the PDB ID is incomplete (e.g. 1A)
+
+        dir : str
+            The directory to store the PDB files in.
+
+        Corner Cases
```
Comment (chuckie82) [inline DeepProjection/pdbGenerator.py:64]: good job thinking about corner cases.
Add these tests to the /tests folder and run pytest and see if they pass

**Review by chuckie82 — COMMENTED** [review]
Adding tests for the new function will improve codecov coverage value. Let's keep it above 90%

File: DeepProjection/tests/test_pdbGenerator.py [hunk 5]
```diff
@@ -13,4 +13,182 @@ def test_getting_pdb(empty_pdbgenerator):
...
+def test_download_pdb_in_range(empty_pdbgenerator):
+    '''download_pdb_in_range() can be run using an "empty" PDBGenerator'''
+    '''download_pdb_in_range() returns True when it encounters no input errors; does not mean it successfully downloaded a file'''
 
+    """
+    Default values
+        1. If upper == None, then range is from [lower, 9ZZZ].
+        2. If dir == None, then download PDB files to current working directory.
+    """
+    # There is no PDB with an ID of '9zzz', so it will be used in default value test.
```
Comment (chuckie82) [inline DeepProjection/tests/test_pdbGenerator.py]: 9zzz may appear one day and this would fail. I think a better check would be a pdb starting with zero, i.e. 0aaa

File: DeepProjection/tests/test_pdbGenerator.py [hunk 6]
```diff
@@ -13,4 +13,182 @@ def test_getting_pdb(empty_pdbgenerator):
...
+
+    # Testing Input Error #6: The first character is '0'; not proper PDB format.
+    assert empty_pdbgenerator.download_pdb_in_range('0Z00', '0Z01', None) == False
+    assert empty_pdbgenerator.download_pdb_in_range('0Z00', '0Z01', 'FAKEDIR') == False
+
+    assert empty_pdbgenerator.download_pdb_in_range('7Zz0', '0Zz1', None) == False
+    assert empty_pdbgenerator.download_pdb_in_range('7Zz0', '0Zz1', 'FAKEDIR') == False
+
+    assert empty_pdbgenerator.download_pdb_in_range('0AAY', '7AAZ', None) == False
+    assert empty_pdbgenerator.download_pdb_in_range('0AAY', '7AAZ', 'FAKEDIR') == False
```
Comment (chuckie82) [inline DeepProjection/tests/test_pdbGenerator.py]: there are too many tests here. Once you test for upper and lower cases, there's no need to keep on testing this
Same with 'FAKEDIR' test

let's only do 1 or 2 per input error type.

**Comment by chuckie82** [comment]
Looking at pypdb code, there's seems to be an API for getting the scores as well:
https://github.com/williamgilpin/pypdb/blob/8bc4aa3b12cc5c0b5e6e7ee6f51b20875324e04c/pypdb/clients/search/search_client.py#L199
